File: src/web/backend/app/schemas/diagnosis.py
"""
诊断 Pydantic 模式
定义诊断数据的请求和响应格式
"""
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field


class DiagnosisBase(BaseModel):
    """诊断基础模式"""
    symptoms: str = Field(default="", max_length=2000, description="症状描述")


class DiagnosisCreate(DiagnosisBase):
    """诊断创建请求模式"""
    disease_id: Optional[int] = Field(None, description="疾病 ID")


class DiagnosisUpdate(BaseModel):
    """诊断更新请求模式"""
    diagnosis_result: Optional[str] = Field(None, description="诊断结果")
    confidence: Optional[float] = Field(None, ge=0, le=1, description="置信度")
    suggestions: Optional[List[str]] = Field(None, description="建议")
    status: Optional[str] = Field(None, description="状态")


class DiagnosisResponse(DiagnosisBase):
    """诊断响应模式"""
    id: int
    user_id: int
    disease_id: Optional[int] = None
    diagnosis_result: Optional[str] = None
    confidence: Optional[float] = None
    suggestions: Optional[List[str]] = None
    status: str
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class DiagnosisWithDisease(DiagnosisResponse):
    """带疾病信息的诊断响应"""
    disease_name: Optional[str] = None
    disease_category: Optional[str] = None


class DiseaseConfidence(BaseModel):
    """病害置信度模型（统一格式，支持多候选病害）"""
    disease_name: str = Field(..., description="病害名称")
    confidence: float = Field(..., ge=0, le=1, description="置信度 0-1")
    disease_class: Optional[int] = Field(None, description="类别ID")


class DiagnosisResult(BaseModel):
    """诊断结果模式（用于 AI 诊断返回）"""
    disease_name: str = Field(..., description="病害名称")
    confidence: float = Field(..., ge=0, le=1, description="置信度")
    severity: Optional[str] = Field(None, description="严重程度")
    description: Optional[str] = Field(None, description="诊断描述")
    recommendations: Optional[str] = Field(None, description="防治建议")
    knowledge_links: List[str] = Field(default_factory=list, description="相关知识链接")


class DiagnosisCreateResponse(BaseModel):
    """诊断创建响应模式"""
    diagnosis_id: str
    disease_name: str
    confidence: float
    confidences: List[DiseaseConfidence] = Field(default_factory=list, description="所有候选病害置信度列表（统一格式）")
    severity: Optional[str] = None
    description: Optional[str] = None
    recommendations: Optional[str] = None
    knowledge_links: List[str] = Field(default_factory=list, description="相关知识链接")
    created_at: datetime


class DiagnosisListResponse(BaseModel):
    """诊断记录列表响应模式（带分页信息）"""
    records: List[DiagnosisResponse]
    total: int
    page: int = Field(1, description="当前页码")
    page_size: int = Field(20, description="每页数量")
    total_pages: int = Field(0, description="总页数")

File: src/web/frontend/src/api/diagnosis.ts
/**
 * 诊断 API 调用函数
 * 负责调用后端诊断相关接口
 */
import { http } from '@/utils/request'

// API 基础路径
const API_BASE = '/diagnosis'

/**
 * 诊断结果类型
 */
export interface DiagnosisResult {
  diagnosis_id: string
  disease_name: string
  confidence: number
  severity?: string
  description?: string
  recommendations?: string
  knowledge_links?: string[]
  created_at: string
}

/**
 * 诊断记录列表响应类型
 */
export interface DiagnosisListResponse {
  records: DiagnosisRecordItem[]
  total: number
  skip: number
  limit: number
}

/**
 * 诊断记录项类型（后端返回格式）
 */
export interface DiagnosisRecordItem {
  id: number
  user_id: number
  disease_id?: number
  symptoms: string
  diagnosis_result?: string
  confidence?: number
  suggestions?: string
  status: string
  created_at: string
  updated_at: string
  disease_name?: string
}

/**
 * 图像诊断请求参数
 */
export interface DiagnosisImageRequest {
  image: File
  symptoms?: string
}

/**
 * 文本诊断请求参数
 */
export interface DiagnosisTextRequest {
  text: string
}

/**
 * 图像诊断
 * @param data 诊断参数（包含图像文件和可选症状描述）
 * @returns 诊断结果
 */
export async function diagnoseImage(data: DiagnosisImageRequest): Promise<DiagnosisResult> {
  const formData = new FormData()
  formData.append('image', data.image)
  if (data.symptoms) {
    formData.append('symptoms', data.symptoms)
  }
  
  return http.post<DiagnosisResult>(`${API_BASE}/image`, formData)
}

/**
 * 文本诊断
 * @param data 文本诊断参数
 * @returns 诊断结果
 */
export async function diagnoseText(data: DiagnosisTextRequest): Promise<DiagnosisResult> {
  return http.post<DiagnosisResult>(`${API_BASE}/text`, data)
}

/**
 * 获取诊断记录列表
 * @param skip 跳过记录数
 * @param limit 返回记录数
 * @returns 诊断记录列表
 */
export async function getDiagnosisRecords(
  skip: number = 0,
  limit: number = 20
): Promise<{ records: DiagnosisRecordItem[]; total: number }> {
  const response = await http.get<DiagnosisListResponse>(`${API_BASE}/records`, {
    params: { skip, limit }
  })
  return {
    records: response.records,
    total: response.total
  }
}

/**
 * 获取诊断详情
 * @param id 诊断记录 ID
 * @returns 诊断详情
 */
export async function getDiagnosisDetail(id: string): Promise<DiagnosisResult> {
  return http.get<DiagnosisResult>(`${API_BASE}/${id}`)
}

/**
 * 更新诊断记录
 * @param id 诊断记录 ID
 * @param data 更新数据
 * @returns 更新后的诊断记录
 */
export async function updateDiagnosis(
  id: string,
  data: Partial<DiagnosisResult>
): Promise<DiagnosisResult> {
  return http.put<DiagnosisResult>(`${API_BASE}/${id}`, data)
}

/**
 * 删除诊断记录
 * @param id 诊断记录 ID
 */
export async function deleteDiagnosis(id: string): Promise<void> {
  return http.delete<void>(`${API_BASE}/${id}`)
}

/**
 * 批量诊断结果项类型
 */
export interface BatchDiagnosisItem {
  index: number
  filename: string
  success: boolean
  diagnosis?: any
  error?: string
  cache_hit?: boolean
}

/**
 * 批量诊断响应类型
 */
export interface BatchDiagnosisResponse {
  success: boolean
  summary: {
    total_images: number
    success_count: number
    failed_count: number
    cache_hits: number
    cache_hit_rate: number
    success_rate: number
  }
  results: BatchDiagnosisItem[]
  performance: {
    total_time_ms: number
    avg_time_per_image_ms: number
  }
  message: string
}

/**
 * 批量图像诊断
 * @param images 图像文件列表（最多10张）
 * @param symptoms 症状描述
 * @param thinkingMode 是否启用 Thinking 推理链
 * @param useGraphRag 是否使用 GraphRAG
 * @returns 批量诊断结果
 */
export async function diagnoseBatch(
  images: File[],
  symptoms: string = '',
  thinkingMode: boolean = false,
  useGraphRag: boolean = false
): Promise<BatchDiagnosisResponse> {
  const formData = new FormData()
  images.forEach((image) => {
    formData.append('images', image)
  })
  if (symptoms) {
    formData.append('symptoms', symptoms)
  }
  formData.append('thinking_mode', String(thinkingMode))
  formData.append('use_graph_rag', String(useGraphRag))
  formData.append('use_cache', 'true')

  return http.post<BatchDiagnosisResponse>(`${API_BASE}/batch`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
    timeout: 600000
  })
}
